### PR TITLE
feat: add a depends-on conditions

### PR DESCRIPTION
**Pull request by sileht**
Fixes MRGFY-455

### CODE REVIEW

File: docs/source/actions/merge.rst [hunk 1]
```diff
@@ -117,6 +120,26 @@ Options
          itself as the commit message. The pull request number will be added to
          end of the title.
 
+.. _depends-on:
+
+⛓️ Defining the pull request dependencies
```
Comment (jd) [inline docs/source/actions/merge.rst:125]:
```suggestion
⛓️ Defining Pull Request Dependencies
```

File: docs/source/actions/merge.rst [hunk 2]
```diff
@@ -117,6 +120,26 @@ Options
...
+----------------------------------------
+
+|premium plan tag|
```
Comment (jd) [inline docs/source/actions/merge.rst:128]:
```suggestion
|open source plan tag|
|premium plan tag|
```

File: docs/source/actions/merge.rst [hunk 3]
```diff
@@ -117,6 +120,26 @@ Options
...
+
+Mergify can wait for other pull requests to be merged before merging this one.
+
+To do this you must add a ``Depends-On:`` lines to you pull request body:
```
Comment (jd) [inline docs/source/actions/merge.rst:132]:
```suggestion
You can specify dependencies between pull requests from the same repository. Mergify waits for the linked pull requests to be merged before merging any pull request with a ``Depends-on:`` header.

To use this feature, adds the ``Depends-On:`` header to the body of your pull request:
```

File: docs/source/actions/queue.rst [hunk 4]
```diff
@@ -16,6 +16,16 @@ Merge queues prevent merging broken pull requests by serializing their merge.
...
 merged in their base branch.
 
+Mergify always respects the `branch protection`_ settings. When the conditions
+match and the ``queue`` action runs, Mergify waits for the branch protection to
+be validated before embarking and merging the pull request.
+
+.. _`branch protection`: https://docs.github.com/en/github/administering-a-repository/about-protected-branches
+
+It can also wait for other pull request to get merged first (see
+:ref:`depends-on`).
```
Comment (jd) [inline docs/source/actions/queue.rst:26]:
```suggestion
Mergify also waits for dependent pull requests to get merged first (see :ref:`depends-on`).
```

File: docs/source/actions/queue.rst [hunk 5]
```diff
@@ -374,4 +384,24 @@ three enqueued pull requests are mergeable.
...
+To do this you must add a ``Depends-On:`` lines to you pull request body:
+
+```
+New awesome feature 🎉
+
+To get the full picture, you may need to look at these pull requests:
+
+Depends-On: #42
+Depends-On: #123
+```
```
Comment (jd) [inline docs/source/actions/queue.rst:405]: same as above

File: mergify_engine/context.py [hunk 6]
```diff
@@ -822,9 +823,33 @@ async def _get_consolidated_data(self, name):
                 for ctext, state in (await self.checks).items()
                 if state == "stale"
             ]
+        elif name == "depends-on":
+            # TODO(sileht): This is the list of merged pull request that
+            # required-by this pull request
```
Comment (jd) [inline mergify_engine/context.py:828]: That sounds wrong no? That's the list of pull request that this PR requires, right?

File: mergify_engine/rules/__init__.py [hunk 7]
```diff
@@ -68,6 +69,30 @@ async def get_branch_protection_conditions(
...
+    ctxt: context.Context,
+) -> typing.List[filter.Filter]:
+    conds: typing.List[filter.Filter] = []
+    for pull_request_number in ctxt.get_depends_on():
+        try:
+            dep_ctxt = await ctxt.repository.get_pull_request_context(
+                pull_request_number
+            )
+        except http.HTTPNotFound:
+            description = f"⛓️ *pull request not found* #{pull_request_number}"
```
Comment (jd) [inline mergify_engine/rules/__init__.py:82]:
```suggestion
            description = f"⛓️  ⚠️ *pull request not found* #{pull_request_number}"
```
or something like that? (not sure how it renders)

File: mergify_engine/rules/__init__.py [hunk 8]
```diff
@@ -68,6 +69,30 @@ async def get_branch_protection_conditions(
...
+        else:
+            description = f"⛓️ **{dep_ctxt.pull['title']}** [#{pull_request_number}]({dep_ctxt.pull['html_url']})"
```
Comment (jd) [inline mergify_engine/rules/__init__.py:84]:
```suggestion
            description = f"⛓️ **{dep_ctxt.pull['title']}** ([#{pull_request_number}]({dep_ctxt.pull['html_url']}))"
```
That looks more like the usual GitHub format.
We might need some function to standardize how we insert pull requests in markdown everywhere (e.g. like in the queue summary).

File: mergify_engine/context.py [hunk 9]
```diff
@@ -822,9 +823,33 @@ async def _get_consolidated_data(self, name):
...
+                    )
+                except http.HTTPNotFound:
+                    continue
+                if ctxt.pull["merged"]:
+                    depends_on.append(f"#{pull_request_number}")
+            return depends_on
         else:
             raise PullRequestAttributeError(name)
 
+    DEPENDS_ON = re.compile(r"^ *Depends-On: #(\d+) *$", re.MULTILINE | re.IGNORECASE)
```
Comment (jd) [inline mergify_engine/context.py:845]: We should have unit tests for this.

It should also support URL-based pull requests, e.g:
```
Depends-On: https://github.com/myorg/myrepo/pull/123
```

**Review by jd — DISMISSED** [review]
I'm actually wondering if that works out of the box for cross-repo?
I guess it's not automatically refreshed, but if that's the only thing we can document it.

**Comment by sileht** [comment]
> I'm actually wondering if that works out of the box for cross-repo?

Yes, we may never be aware that the dependency has been merged.

> I guess it's not automatically refreshed, but if that's the only thing we can document it.

File: mergify_engine/context.py (hunk 6 above)
Comment (sileht) [reply mergify_engine/context.py:828]: The auto generates rule conditions have all `Depends-On:` but the consolidated_data only have the merged one.

This way, the condition will match only when the pull request is merged.

Comment (jd) [reply mergify_engine/context.py:828]: Ok, we're saying the same thing, it's just the comment format make me thought you meant the other way around.
```suggestion
            # This is the list of merged pull requests that are required by this pull request
```

**Comment by jd** [comment]
I'm actually wondering if that works out of the box for cross-repo?

Yes, we may never be aware that the dependency has been merged.

Ok, that's really not a big deal. We can improve this later for repo that are both using Mergify. For now we can just document it works but you need to `refresh` 🤷

**Comment by sileht** [comment]
> I'm actually wondering if that works out of the box for cross-repo?
> 
> Yes, we may never be aware that the dependency has been merged.
> 
> Ok, that's really not a big deal. We can improve this later for repo that are both using Mergify. For now we can just document it works but you need to `refresh` shrug

It doesn't work yet, they are additional work to do to fetch the pull request state of other repositories.

I will do this in another pull request because I need to implement it for other repositories.

File: docs/source/actions/merge.rst [hunk 10]
```diff
@@ -117,6 +119,33 @@ Options
...
+
+⛓️ Defining Pull Request Dependencies
+-------------------------------------
+
+|premium plan tag|
+|open source plan tag|
+
+You can specify dependencies between pull requests from the same repository.
+Mergify waits for the linked pull requests to be merged before merging any pull
+request with a ``Depends-on:`` header.
```
Comment (jd) [inline docs/source/actions/merge.rst:132]:
```suggestion
request with a ``Depends-on:`` header.
```
```suggestion
request with a ``Depends-On:`` header.
```

File: docs/source/actions/merge.rst [hunk 11]
```diff
@@ -117,6 +119,33 @@ Options
...
+
+    New awesome feature 🎉
+
+    To get the full picture, you may need to look at these pull requests:
+
+    Depends-On: #42
+    Depends-On: https://github.com/organization/repository/pull/123
+
+
+This feature does not work for cross repository dependencies.
```
Comment (jd) [inline docs/source/actions/merge.rst:147]:
```suggestion
.. warning::

    This feature does not work for cross-repository dependencies.
```

File: docs/source/actions/queue.rst [hunk 12]
```diff
@@ -374,4 +383,30 @@ three enqueued pull requests are mergeable.
...
+
+⛓️ Defining Pull Request Dependencies
+-------------------------------------
+
+|premium plan tag|
+|open source plan tag|
+
+You can specify dependencies between pull requests from the same repository.
+Mergify waits for the linked pull requests to be merged before merging any pull
+request with a ``Depends-on:`` header.
```
Comment (jd) [inline docs/source/actions/queue.rst:396]:
```suggestion
request with a ``Depends-On:`` header.
```

File: docs/source/global-substitutions.rst [hunk 13]
```diff
@@ -12,3 +12,7 @@
 .. |beta tag| raw:: html
 
    <div class="feature-tag"><span class="badge badge-info">Beta Feature ⚙️</span></div>
+
+.. |open source plan tag| raw:: html
+
+   <div class="feature-tag"><span class="badge badge-info">Open Source<br /> Feature 💖</span></div>
```
Comment (jd) [inline docs/source/global-substitutions.rst:18]:
```suggestion
   <div class="feature-tag"><span class="badge badge-primary">Open Source<br /> Feature 💖</span></div>
```

**Review by jd — CHANGES_REQUESTED** [review]
small typo but 👍

File: docs/source/actions/merge.rst [hunk 14]
```diff
@@ -117,6 +119,35 @@ Options
...
+    New awesome feature 🎉
+
+    To get the full picture, you may need to look at these pull requests:
+
+    Depends-On: #42
+    Depends-On: https://github.com/organization/repository/pull/123
+
+.. warning::
+
+    This feature does not work for cross repository dependencies.
```
Comment (jd) [inline docs/source/actions/merge.rst:148]:
```suggestion
    This feature does not work for cross-repository dependencies.
```

**Review by jd — APPROVED** [review]

**Pull request merged**